CLN: move top-level dirs

jreback opened the pull request:
xref #13634

- [x] ``from pandas.types.concat import union_categoricals`` ?

- ``pandas.types`` -> ``pandas.core.typed``
- ``pandas.computation`` -> ``pandas.core.computation``
- ``pandas.sparse`` -> ``pandas.core.sparse``

jreback committed:
CLN: move pandas.types -> pandas.core.typed

xref #13634

jreback commented:
this was probably the one with the most actual code changes.

jreback commented on doc/source/whatsnew/v0.20.0.txt:
```diff
@@ -1000,12 +1000,13 @@ New Behavior:
-^^^^^^^^^^^^^^^^^^^^^^^
+Modules Privacy Has Changed
+^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
```
we can make this more prominent later.

jreback committed:
CLN: move pandas.computation -> pandas.core.computation

jreback commented:
``from pandas.types.concat import union_categoricals``
we need to change this.

``from pandas.api import union_categoricals``?

also should change ``infer_dtype`` to come from ``pandas.api`` as well? (xref #15936) (I will do as a followup I think).

jreback commented:
cc @chris-b1 @jorisvandenbossche

chris-b1 commented:
I suppose `union_categoricals` and `infer_dtype` should go in `pandas.api.types`?

jreback commented:
yep I think that's right place.

jreback committed:
move pandas.api.lib.infer_dtype -> pandas.api.types

move union_categoricals to pandas.api.types

make feather single test

fix odd tz hashing issue

remove circular import

pandas.core.common warning message

jorisvandenbossche commented:
Is there a reason you use `core.typed` instead of `core.types` ?

jreback commented:
> Is there a reason you use core.typed instead of core.types ?

yes, unfortunately in python 2 if you do ``import types`` and you are *in*  the ``pandas.core`` directly you get ``pandas.core.types``, rather than the built in module (it get shadowed).

we use this in ``pandas.core.series``. I *can* get around this, but simpler to just rename.

jorisvandenbossche commented:
Ah, just saw you commit message that mentioned it as well. 
I might prefer 'dtypes' over 'typed', but not that important

jreback committed:
rename .typed -> .dtypes

jreback committed:
move privacy changes to development section

jreback committed:
CLN: pandas.sparse -> pandas.core.sparse

jreback commented:
renamed ``typed`` -> ``dtypes``, so a bit more natural now

``pandas.core.dtypes``

jreback committed:
try with pyargs

jreback committed:
pep

jreback committed:
more verbose built test

jreback committed:
moar ci

jreback committed:
move fixtures to top-level conftest

jreback commented:
finally got this bad boy working. apparently ``pytest``doesn't like nested conftests.py (of course *only* when you are running using ``pytest.main()``; it works just fine for development.

jorisvandenbossche commented:
Great!
